File: client/src/data/curriculum/python_03_functions.ts
import type { DayTrack, Challenge, ConceptPrimerData, TestCase } from '../../types';

export const DAY03_TRACK: DayTrack = {
  partNumber: 3,
  partId: 3,
  dayNumber: 3,
  id: 3,
  title: 'Part 3: Functions, Scopes, Closures & First-Class Citizens',
  subtitle: 'Argument unpacking, mutable defaults, LEGB scope resolution, and closure factories',
  description: 'Master the mechanics of Python functions as first-class objects. Dissect stack frames, understand lexical scope via the LEGB rule, master `*args` and `**kwargs`, avoid the dangerous mutable default argument pitfall, and build configurable pipeline composers and memoizing closures.',
  iconName: 'Zap',
  badge: 'Part 3 • Functions & Scopes',
  libraryMechanics: {
    libraryName: 'Python Execution Frames & Closures',
    tagline: 'Stack frames, lexical scopes, closure cell objects, and first-class function dispatch.',
    overview: `### ⚡ Functions are First-Class Objects
In Python, functions are runtime objects of type \`function\`. They can be:
- Passed as arguments to other functions (\`map\`, \`filter\`, \`sorted\`).
- Returned dynamically from functions (higher-order functions / function factories).
- Assigned to variables, placed in lists, and stored in dictionaries.
- Inspected at runtime via attributes like \`__name__\`, \`__code__\`, \`__defaults__\`, and \`__closure__\`.

### 🔍 The LEGB Scope Resolution Rule
When Python evaluates a variable name inside a function, it traverses four lexical scopes in strict order:
1. **Local (L):** Names defined inside the currently executing function frame.
2. **Enclosing (E):** Names in enclosing nested function definitions (closures).
3. **Global (G):** Names defined at module-level scope (\`__main__\` or file top-level).
4. **Built-in (B):** Preloaded Python identifiers (\`len\`, \`range\`, \`ValueError\`, \`print\`).`,
    whyItExists: `Functions are the atomic units of logic and abstraction:
- **Closures:** A function that retains access to variables from its enclosing scope even after that outer function has finished executing. CPython wraps these in \`cell\` objects.
- **Parametric Flexibility:** Combining positional-only (\`/\`), keyword-only (\`*\`), variable positional (\`*args\`), and keyword arguments (\`**kwargs\`) enables authoring extensible APIs.
- **No Hidden Re-evaluation:** Default arguments in Python are evaluated **once** at function definition time, NOT at every call! Understanding this prevents catastrophic bugs with mutable defaults like \`def f(x=[])\`.`,
    coreAnatomy: {
      objectName: 'CPython PyFrameObject & PyFunctionObject',
      description: 'The internal state machine for function execution and closure variables.',
      fields: [
        {
          name: 'co_varnames',
          type: 'tuple[str]',
          role: 'Names of local variables and arguments.'
        },
        {
          name: 'co_freevars',
          type: 'tuple[str]',
          role: 'Names of variables captured from enclosing scopes (closure variables).'
        },
        {
          name: '__closure__',
          type: 'tuple[cell]',
          role: 'Tuple of cell objects holding pointers to captured lexical variables.'
        },
        {
          name: '__defaults__',
          type: 'tuple[Any]',
          role: 'Evaluated default argument objects preserved on the function object.'
        }
      ],
      memoryDiagramAscii: `+-------------------------------------------------------------+
|               Python Closure & Cell Memory Model            |
|                                                             |
|  Outer Function Frame (Closed)                              |
|  +---------------------------+                              |
|  | Cell Object (0xCAFE)      | <---+                        |
|  | payload: count = 0        |     |                        |
|  +---------------------------+     |                        |
|                                    |                        |
|  Inner Returned Function Object    |                        |
|  +-------------------------------+ |                        |
|  | __closure__ = (Cell@0xCAFE,) -+ |                        |
|  | f() -> mutates Cell@0xCAFE      |                        |
|  +-------------------------------+                          |
+-------------------------------------------------------------+`
    },
    chapters: [
      {
        id: 'ch1-args-kwargs',
        title: 'Modern Argument Signatures: *, /, *args, **kwargs',
        icon: 'Layers',
        summary: 'Control parameter semantics with positional-only and keyword-only boundaries.',
        markdownContent: `### Strict Signature Boundaries

Python 3.8+ introduced full control over argument binding:

\`\`\`python
def configure(host, port=8080, /, debug=False, *, timeout=30, **extra):
    pass
\`\`\`

- **Before \`/\`**: **Positional-only**. Callers *must* pass \`host\` and \`port\` positionally (cannot write \`host="localhost"\`).
- **Between \`/\` and \`*\`**: Positional or keyword.
- **After \`*\`**: **Keyword-only**. Callers *must* pass \`timeout=...\` explicitly.
- **\`*args\`**: Captures any excess positional arguments as a \`tuple\`.
- **\`**kwargs\`**: Captures any excess keyword arguments as a \`dict\`.`,
        codeSnippets: [
          {
            id: 'snip-args-kwargs',
            title: 'Variable Arguments Forwarding',
            code: `def logger(prefix, *args, **kwargs):
    print(f"[{prefix.upper()}]", *args)
    if kwargs:
        print(" Extra:", kwargs)

logger("audit", "User logged in", "IP: 127.0.0.1", user_id=42)`,
            explanation: '*args captures positional arguments as a tuple; **kwargs captures keyword arguments as a dict.'
          }
        ]
      },
      {
        id: 'ch2-closures',
        title: 'Closures and the `nonlocal` Keyword',
        icon: 'Sparkles',
        summary: 'Enclosing scopes and state encapsulation without classes.',
        markdownContent: `### Encapsulating State in Closures

A closure allows creating lightweight stateful objects using pure functions:

\`\`\`python
def make_counter(start: int = 0):
    count = start  # Enclosed variable in cell
    
    def increment():
        nonlocal count  # Declares that count lives in enclosing scope
        count += 1
        return count
        
    return increment

counter = make_counter(10)
print(counter())  # 11
print(counter())  # 12
\`\`\`

Without \`nonlocal\`, assigning \`count += 1\` would cause Python to treat \`count\` as an uninitialized local variable, raising \`UnboundLocalError\`!`,
        codeSnippets: [
          {
            id: 'snip-closure-inspection',
            title: 'Inspecting Closure Cells',
            code: `def factory(factor):
    return lambda x: x * factor

times3 = factory(3)
print("times3(10):", times3(10))
print("Cell content:", times3.__closure__[0].cell_contents)`,
            explanation: 'Functions carry a __closure__ tuple of cell objects retaining variables from enclosing lexical scopes.'
          }
        ]
      }
    ],
    commonTraps: [
      {
        title: 'The Mutable Default Parameter Trap',
        badSnippet: `def append_item(item, target_list=[]):\n    target_list.append(item)\n    return target_list`,
        badExplanation: 'Default argument expressions are evaluated once at function definition time, making mutable defaults persistent across all invocations.',
        goodSnippet: `def append_item(item, target_list=None):\n    if target_list is None:\n        target_list = []\n    target_list.append(item)\n    return target_list`,
        goodExplanation: 'Use None as default sentinel and instantiate a fresh list inside the function call.',
        perfImpact: 'Prevents insidious cross-invocation memory corruption and unintended data leaks.'
      }
    ],
    apiCheatSheet: [
      {
        name: 'partial',
        category: 'Higher-Order',
        signature: 'functools.partial(func, /, *args, **keywords)',
        summary: 'Returns a new partial object which when called behaves like func called with positional and keyword arguments pre-filled.',
        parameters: [
          { name: 'func', type: 'callable', desc: 'Function to partially apply.' },
          { name: '*args', type: 'tuple', desc: 'Pre-filled positional arguments.' }
        ],
        returns: 'Partial callable object.',
        exampleSnippet: 'base2 = partial(int, base=2)'
      },
      {
        name: 'callable',
        category: 'Inspection',
        signature: 'callable(obj) -> bool',
        summary: 'Returns True if the object appears callable (functions, methods, classes, objects with __call__).',
        parameters: [
          { name: 'obj', type: 'object', desc: 'Object to test.' }
        ],
        returns: 'bool indicating if callable.',
        exampleSnippet: 'if callable(handler): handler()'
      },
      {
        name: 'wraps',
        category: 'Decorators',
        signature: 'functools.wraps(wrapped)',
        summary: 'Decorator factory that copies __name__, __doc__, and other introspection metadata to wrapper functions.',
        parameters: [
          { name: 'wrapped', type: 'callable', desc: 'Original target function.' }
        ],
        returns: 'Decorator updating wrapper metadata.',
        exampleSnippet: '@wraps(func)'
      }
    ],
    interactiveWidgetType: 'python-memory'
  },
  challenges: [
    {
      id: 'python-p3-c1',
      dayId: 3,
      partId: 3,
      title: 'Configurable Pipeline Function Composer',
      slug: 'configurable-pipeline-function-composer',
      difficulty: 'Intermediate',
      category: 'Higher-Order Functions',
      summary: 'Compose an arbitrary sequence of callable transformation stages into a single callable execution pipeline with optional debugging traces.',
      estimatedTime: '20 min',
      hints: [
        'Verify that the variable positional arguments sequence is non-empty, and check that each supplied function satisfies Python\'s callable interface.',
        'Construct an inner closure that receives the initial argument and iterates sequentially through each function, updating an accumulator with each stage\'s result.',
        'When debug mode is requested, track execution by recording the zero-based step index, resolving the callable name (with fallback handling for anonymous or callable objects), and storing the stage output.',
        'Wrap each stage invocation in an exception handler and use explicit exception chaining to preserve the original exception while raising a informative RuntimeError.'
      ],
      instructions: `Write a function \`compose_pipeline(*funcs: callable, debug: bool = False) -> callable\` that:
1. **Validation**:
   - If no functions are provided, raise a \`ValueError\` indicating that at least one function is required.
   - Ensure every argument provided is callable; if any is not, raise a \`TypeError\`.
2. **Composed Function Execution**:
   - The returned pipeline function must accept an initial input value.
   - It executes the functions sequentially from left to right, threading the output of each stage as the input to the next.
   - If any stage raises an exception during execution, catch it and raise a \`RuntimeError\` with the original exception explicitly chained. The error message should convey the failed stage index, the function's name (with a fallback name if unnamed), and the underlying error message.
3. **Debug Mode**:
   - If \`debug=False\` (default): Returns the final transformed output value directly.
   - If \`debug=True\`: Returns a dictionary containing:
     - \`"result"\`: The final transformed output value.
     - \`"trace"\`: A list of dictionaries documenting each step, where each entry contains:
       - \`"step"\`: The zero-based integer index of the stage.
       - \`"func"\`: The string identifier/name of the function.
       - \`"output"\`: The intermediate value produced by that stage.`,
      starterCode: `def compose_pipeline(*funcs: callable, debug: bool = False) -> callable:
    """
    Compose multiple callable functions into a single left-to-right execution pipeline.
    
    Args:
        *funcs: Variable number of callable transformation functions.
        debug: If True, returned pipeline outputs an execution trace.
        
    Returns:
        Composed callable function.
    """
    # TODO: Validate functions, return composed callable with optional trace
    pass
`,
      solutionCode: `def compose_pipeline(*funcs: callable, debug: bool = False) -> callable:
    if not funcs:
        raise ValueError("At least one function must be provided")
        
    for f in funcs:
        if not callable(f):
            raise TypeError("All pipeline stages must be callable")
            
    def pipeline(initial_val):
        current = initial_val
        trace = []
        
        for idx, f in enumerate(funcs):
            fn_name = getattr(f, "__name__", f"stage_{idx}")
            try:
                current = f(current)
            except Exception as e:
                raise RuntimeError(f"Pipeline failed at stage {idx} ({fn_name}): {e}") from e
                
            if debug:
                trace.append({
                    "step": idx,
                    "func": fn_name,
                    "output": current,
                })
                
        if debug:
            return {"result": current, "trace": trace}
        return current
        
    return pipeline
`,
      testCases: [
        {
          id: 't1',
          name: 'Standard Three-Stage Pipeline',
          inputDescription: 'funcs=[str.strip, str.lower, len], val="  Hello World  ", debug=False',
          expectedOutput: '11 (length of "hello world")'
        },
        {
          id: 't2',
          name: 'Debug Trace Mode',
          inputDescription: 'funcs=[lambda x: x + 1, lambda x: x * 2], val=5, debug=True',
          expectedOutput: "{result: 12, trace: [{step: 0, func: '<lambda>', output: 6}, {step: 1, func: '<lambda>', output: 12}]}"
        },
        {
          id: 't3',
          name: 'Pipeline Stage Failure Handling',
          inputDescription: 'funcs=[lambda x: x / 0], val=10',
          expectedOutput: 'Raises RuntimeError with stage index and original error'
        }
      ],
      benchmarkTargetMs: 0.05,
      memoryTargetMb: 0.1,
      conceptPrimer: {
        title: 'Function Composition & Pipeline Architecture',
        subtitle: 'Chaining pure transformations into clean, testable data workflows',
        overview: 'Pipelines in machine learning, data engineering, and web middleware are built by composing small pure functions.',
        mentalModel5s: 'Loop through `funcs`, updating `val = f(val)` -> Wrap in a try/except for diagnostic traceability.',
        visualAnalogy: 'An assembly line where each worker applies a specialized polish before handing the object to the next station.',
        pitfalls: [
          'Forgetting that `getattr(f, "__name__", ...)` is needed for anonymous lambdas or callable objects.',
          'Not chaining exceptions using `from e`.'
        ],
        progressiveHints: [
          'Step 1: Check `if not funcs:` and verify `all(callable(f) for f in funcs)`.',
          'Step 2: Define the inner `pipeline(initial_val)` closure.',
          'Step 3: Accumulate steps in a `trace` list if `debug=True`.',
          'Step 4: Return `pipeline` callable.'
        ],
        mathFormulas: [
          {
            title: 'Function Composition Pipeline',
            latex: '(f_n \\circ \\dots \\circ f_2 \\circ f_1)(x) = f_n(\\dots(f_2(f_1(x))))',
            explanation: 'Sequential unary composition where the output of each function feeds into the next.'
          }
        ],
        naiveVsIdiomatic: {
          naiveCode: `# Hardcoded nested function calls
res = f3(f2(f1(x)))`,
          naiveExplanation: 'Inflexible, unconfigurable, and cannot dynamically add or remove processing stages.',
          idiomaticCode: `pipeline = compose_pipeline(f1, f2, f3)
res = pipeline(x)`,
          idiomaticExplanation: 'Declarative, dynamic, reusable, and self-documenting.',
          speedupText: 'Infinite architectural flexibility'
        },
        memoryLayout: {
          title: 'Stack Frame Unwinding',
          content: 'The composed closure maintains a reference to the `funcs` tuple in its `__closure__` cell.',
          diagramAscii: `pipeline.__closure__[0] -> tuple of [f1, f2, f3]`,
          keyRule: 'Captured functions are never copied; only pointer references are retained in the closure.'
        },
        keyTakeaways: [
          'Functions in Python can be composed dynamically.',
          'Use `callable()` to validate parameters.',
          'Provide debug instrumentation for observable execution in production systems.'
        ]
      }
    },
    {
      id: 'python-p3-c2',
      dayId: 3,
      partId: 3,
      title: 'State-Preserving Bounded Memoizer Closure',
      slug: 'state-preserving-bounded-memoizer-closure',
      difficulty: 'Intermediate',
      category: 'Closures & Cache Design',
      summary: 'Build a closure-based function decorator that caches expensive computation results with FIFO eviction and statistics tracking.',
      estimatedTime: '20 min',
      hints: [
        'Validate `isinstance(maxsize, int) and maxsize > 0`.',
        'Inside `make_bounded_memoizer`, declare `cache = {}`, `hits = 0`, `misses = 0`.',
        'In wrapper(*args), check `if args in cache: hits += 1; return cache[args]`.',
        'If evicting, use `del cache[next(iter(cache))]` to remove the oldest FIFO entry.',
        'Attach `cache_info` and `cache_clear` to wrapper, and decorate with `@wraps(fn)`.'
      ],
      instructions: `Write a function \`make_bounded_memoizer(maxsize: int = 128) -> callable\` that:
1. **Validation**:
   - If \`not isinstance(maxsize, int) or maxsize <= 0\`, raise \`ValueError("maxsize must be a positive integer")\`.
2. **Decorator Factory**:
   - Returns a decorator function \`memoize(fn: callable) -> callable\`.
   - The returned wrapped function caches results based on the tuple of arguments \`(*args,)\` passed to \`fn\`.
3. **Eviction Policy**:
   - If a new computation must be cached and the cache size has reached \`maxsize\`, evict the **oldest** inserted cache entry (FIFO eviction).
4. **Cache Metrics & Management**:
   - The wrapped function must expose two callable helper attributes:
     - \`.cache_info() -> dict\`: returns \`{"hits": int, "misses": int, "size": int, "maxsize": int}\`.
     - \`.cache_clear() -> None\`: clears the cache dictionary and resets \`hits\` and \`misses\` to \`0\`.
5. **Metadata Preservation**:
   - The wrapper function must preserve the original function's \`__name__\` and \`__doc__\` (using \`functools.wraps\`).`,
      starterCode: `def make_bounded_memoizer(maxsize: int = 128) -> callable:
    """
    Factory creating a memoization decorator with bounded FIFO cache size and metrics.
    
    Args:
        maxsize: Maximum number of results to cache (default 128).
        
    Returns:
        Decorator function.
    """
    # TODO: Implement bounded memoizer closure with cache_info and cache_clear
    pass
`,
      solutionCode: `from functools import wraps

def make_bounded_memoizer(maxsize: int = 128) -> callable:
    if not isinstance(maxsize, int) or isinstance(maxsize, bool) or maxsize <= 0:
        raise ValueError("maxsize must be a positive integer")
        
    def decorator(fn: callable) -> callable:
        if not callable(fn):
            raise TypeError("fn must be callable")
            
        cache = {}
        hits = 0
        misses = 0
        
        @wraps(fn)
        def wrapper(*args):
            nonlocal hits, misses
            key = args
            
            if key in cache:
                hits += 1
                return cache[key]
                
            misses += 1
            result = fn(*args)
            
            if len(cache) >= maxsize:
                # Evict oldest entry (FIFO using dict order in Python 3.7+)
                oldest_key = next(iter(cache))
                del cache[oldest_key]
                
            cache[key] = result
            return result
            
        def cache_info():
            return {
                "hits": hits,
                "misses": misses,
                "size": len(cache),
                "maxsize": maxsize,
            }
            
        def cache_clear():
            nonlocal hits, misses
            cache.clear()
            hits = 0
            misses = 0
            
        wrapper.cache_info = cache_info
        wrapper.cache_clear = cache_clear
        return wrapper
        
    return decorator
`,
      testCases: [
        {
          id: 't1',
          name: 'Fibonacci Memoization & Hit Tracking',
          inputDescription: 'maxsize=10, compute fib(5) multiple times',
          expectedOutput: 'hits=4, misses=6, size=6'
        },
        {
          id: 't2',
          name: 'FIFO Cache Eviction on Capacity Limit',
          inputDescription: 'maxsize=2, insert keys 1, 2, then 3',
          expectedOutput: 'Key 1 evicted; cache size remains 2'
        },
        {
          id: 't3',
          name: 'Cache Clear Functionality',
          inputDescription: 'wrapper.cache_clear()',
          expectedOutput: 'hits=0, misses=0, size=0'
        }
      ],
      benchmarkTargetMs: 0.05,
      memoryTargetMb: 0.1,
      conceptPrimer: {
        title: 'Stateful Closures & Memoization Mechanics',
        subtitle: 'Trading space for time with function caches',
        overview: 'Expensive calculations (dynamic programming, database reads, API lookups) use memoization caches to return previously computed results in $O(1)$ time.',
        mentalModel5s: 'Encapsulate `cache`, `hits`, `misses` inside closure -> Attach `.cache_info` and `.cache_clear` to `wrapper`.',
        visualAnalogy: 'A desk drawer containing answered letters: if the same question arrives, read the drawer answer instead of recalculating.',
        pitfalls: [
          'Forgetting `nonlocal hits, misses` when reassigning counters in wrapper.',
          'Not preserving `__name__` and `__doc__` with `@functools.wraps`.'
        ],
        progressiveHints: [
          'Step 1: Check `maxsize > 0`.',
          'Step 2: Define `cache = {}`, `hits = 0`, `misses = 0` inside the decorator.',
          'Step 3: In `wrapper(*args)`, use `args` as the cache key.',
          'Step 4: If `len(cache) >= maxsize`, pop the oldest key: `del cache[next(iter(cache))]`.',
          'Step 5: Attach `wrapper.cache_info` and `wrapper.cache_clear`.'
        ],
        mathFormulas: [
          {
            title: 'Cache Hit Ratio',
            latex: '\\text{Hit Ratio} = \\frac{\\text{hits}}{\\text{hits} + \\text{misses}}',
            explanation: 'Quantifies cache efficiency and latency savings over repeated invocations.'
          }
        ],
        naiveVsIdiomatic: {
          naiveCode: `# Global cache variable
GLOBAL_CACHE = {}
def calc(x):
    if x in GLOBAL_CACHE: ...`,
          naiveExplanation: 'Pollutes global namespace, prevents multiple isolated instances, and risks concurrency collisions.',
          idiomaticCode: `@make_bounded_memoizer(maxsize=100)
def calc(x): ...`,
          idiomaticExplanation: 'Each decorated function owns an isolated, private closure cache with introspection hooks.',
          speedupText: '100x speedup on recursive algorithms'
        },
        memoryLayout: {
          title: 'Closure Cell References',
          content: 'The wrapper holds cells referencing `cache`, `hits`, `misses`, and `maxsize`.',
          diagramAscii: `wrapper.__closure__ -> [cell(cache), cell(hits), cell(misses), cell(maxsize)]`,
          keyRule: 'Cells remain alive in RAM as long as wrapper exists, independent of outer scope lifecycle.'
        },
        keyTakeaways: [
          'Closures retain private encapsulated state without requiring explicit classes.',
          'Use `next(iter(dict))` to access the oldest inserted key in Python 3.7+.',
          'Always use `@functools.wraps` on decorators to keep introspection intact.'
        ]
      }
    }
  ]
};

export const PYTHON_PART03_TRACK = DAY03_TRACK;
export const testCases = DAY03_TRACK.challenges.flatMap((c) => c.testCases);
export type { DayTrack, Challenge, ConceptPrimerData, TestCase };
export default DAY03_TRACK;

File: server/curriculum/python_03_functions.py
"""
Part 3: Functions, Scopes, Closures & First-Class Citizens
PyMastery Progressive Zero-to-Hero Pure Python Curriculum
"""

from functools import wraps
from typing import Dict, Any, List

DAY_METADATA = {
    "day_id": "python_03",
    "day_number": 3,
    "title": "Part 3: Functions, Scopes, Closures & First-Class Citizens",
    "tagline": "Argument unpacking, mutable defaults, LEGB scope resolution, and closure factories.",
    "estimated_time": "1-2 hours",
    "concepts_covered": [
        "First-class function semantics and higher-order functional patterns",
        "The LEGB lexical scope hierarchy and the `nonlocal` keyword",
        "CPython execution frames, code objects, and closure cell references",
        "Positional-only `/`, keyword-only `*`, `*args`, and `**kwargs` signatures",
        "State encapsulation via closure factories and function memoization"
    ]
}

CONCEPT_PRIMER = r"""# Part 3: Functions, Scopes, Closures & First-Class Citizens

In Python, functions are not merely code blocks; they are **first-class runtime objects**. You can pass them as arguments, return them from other functions, store them in data structures, and introspect their internal bytecode.

---

## 1. The LEGB Scope Resolution Hierarchy

When an identifier is looked up, Python scans four nested scopes:

1. **Local (L):** Assigned directly inside the active function frame (`x = 10`).
2. **Enclosing (E):** Found in the local namespace of any enclosing `def` blocks (closures).
3. **Global (G):** Assigned at module top-level (`__main__`).
4. **Built-in (B):** Preloaded language identifiers (`len`, `range`, `ValueError`).

```python
x = "global"

def outer():
    x = "enclosing"
    def inner():
        # Searches Local -> Enclosing (found!)
        print(x)  # prints 'enclosing'
    inner()
```

---

## 2. Closures & Cell Objects

When an inner function references a variable from an enclosing scope, CPython stores that variable inside a heap-allocated **`cell`** object.

Even after the outer function finishes executing and its frame is discarded, the inner function maintains an active pointer to the `cell` via its `__closure__` tuple.

```python
def make_multiplier(n):
    return lambda x: x * n

double = make_multiplier(2)
# The integer 2 is kept alive inside double.__closure__[0].cell_contents!
print(double(5))  # 10
```

---

## 3. The Dangerous Mutable Default Argument

Default arguments are evaluated **once** at compile/definition time:
```python
def bad_append(val, lst=[]):  # TRAP: lst is created ONCE!
    lst.append(val)
    return lst
```
Instead, always use sentinel defaults:
```python
def safe_append(val, lst=None):
    if lst is None:
        lst = []
    lst.append(val)
    return lst
```
"""

WALKTHROUGH = r"""# Walkthrough: Building a Simple Timer Closure

```python
import time

def make_timer():
    t_start = time.perf_counter()
    
    def elapsed():
        return time.perf_counter() - t_start
        
    return elapsed

timer = make_timer()
time.sleep(0.05)
print(f"Elapsed: {timer():.4f} seconds")
```
"""

# -----------------------------------------------------------------------------
# Challenge 1: Configurable Pipeline Function Composer
# -----------------------------------------------------------------------------

CHALLENGE_1 = {
    "id": "python-p3-c1",
    "title": "Configurable Pipeline Function Composer",
    "difficulty": "Intermediate",
    "category": "Higher-Order Functions",
    "description": (
        "Compose an arbitrary sequence of callable transformation stages into a single callable execution pipeline "
        "with optional debugging traces."
    ),
    "instructions": (
        "Write a function `compose_pipeline(*funcs: callable, debug: bool = False) -> callable` that:\n"
        "1. Validates that at least one function argument is provided; if empty, raises a `ValueError`.\n"
        "2. Ensures every supplied argument is callable; if any is not, raises a `TypeError`.\n"
        "3. Returns a pipeline callable that accepts an initial input value and executes the functions sequentially from left to right, threading the output of each stage into the next.\n"
        "4. If any stage raises an exception, catches it and raises a `RuntimeError` explicitly chained from the original exception, conveying the stage index, function name (with a fallback name if unnamed), and error message.\n"
        "5. If `debug=False` (default), returns the final transformed value directly.\n"
        "6. If `debug=True`, returns a dictionary with 'result' (the final transformed output) and 'trace' (a list of dictionaries documenting each step with 'step' index, 'func' name, and 'output' value)."
    ),
    "starter_code": r'''def compose_pipeline(*funcs: callable, debug: bool = False) -> callable:
    """
    Compose multiple callable functions into a single left-to-right execution pipeline.
    
    Args:
        *funcs: Variable number of callable transformation functions.
        debug: If True, returned pipeline outputs an execution trace.
        
    Returns:
        Composed callable function.
    """
    # TODO: Validate functions, return composed callable with optional trace
    pass
''',
    "reference_solution": r'''def compose_pipeline(*funcs: callable, debug: bool = False) -> callable:
    if not funcs:
        raise ValueError("At least one function must be provided")
        
    for f in funcs:
        if not callable(f):
            raise TypeError("All pipeline stages must be callable")
            
    def pipeline(initial_val):
        current = initial_val
        trace = []
        
        for idx, f in enumerate(funcs):
            fn_name = getattr(f, "__name__", f"stage_{idx}")
            try:
                current = f(current)
            except Exception as e:
                raise RuntimeError(f"Pipeline failed at stage {idx} ({fn_name}): {e}") from e
                
            if debug:
                trace.append({
                    "step": idx,
                    "func": fn_name,
                    "output": current,
                })
                
        if debug:
            return {"result": current, "trace": trace}
        return current
        
    return pipeline
''',
    "test_suite": r'''def run_tests(candidate_func):
    report = {"passed": True, "tests_run": 0, "errors": []}
    
    def assert_test(condition, msg):
        report["tests_run"] += 1
        if not condition:
            report["passed"] = False
            report["errors"].append(msg)
            raise AssertionError(msg)

    # Test 1: Standard composition
    pipe = candidate_func(str.strip, str.lower, len)
    res = pipe("  Hello World  ")
    assert_test(res == 11, f"Expected 11, got {res}")

    # Test 2: Debug mode trace
    pipe_debug = candidate_func(lambda x: x + 1, lambda x: x * 2, debug=True)
    res_debug = pipe_debug(5)
    assert_test(isinstance(res_debug, dict), "Debug output must be a dict")
    assert_test(res_debug["result"] == 12, f"Debug result mismatch: {res_debug['result']}")
    assert_test(len(res_debug["trace"]) == 2, f"Trace length mismatch: {len(res_debug['trace'])}")

    # Test 3: Exception handling in stage
    pipe_err = candidate_func(lambda x: x / 0)
    try:
        pipe_err(10)
        assert_test(False, "Should raise RuntimeError on stage division by zero")
    except RuntimeError:
        report["tests_run"] += 1

    # Test 4: Empty funcs validation
    try:
        candidate_func()
        assert_test(False, "Should raise ValueError for no funcs")
    except ValueError:
        report["tests_run"] += 1

    return report
''',
    "hints": [
        "Verify that the variable positional arguments sequence is non-empty, and check that each supplied function satisfies Python's callable interface.",
        "Construct an inner closure that receives the initial argument and iterates sequentially through each function, updating an accumulator with each stage's result while safely resolving callable names with fallbacks for anonymous lambdas.",
        "Wrap each stage invocation in an exception handler and use explicit exception chaining to preserve the original exception while raising an informative RuntimeError."
    ]
}

# -----------------------------------------------------------------------------
# Challenge 2: State-Preserving Bounded Memoizer Closure
# -----------------------------------------------------------------------------

CHALLENGE_2 = {
    "id": "python-p3-c2",
    "title": "State-Preserving Bounded Memoizer Closure",
    "difficulty": "Intermediate",
    "category": "Closures & Cache Design",
    "description": (
        "Build a closure-based function decorator that caches expensive computation results with FIFO eviction "
        "and statistics tracking."
    ),
    "instructions": (
        "Write a function `make_bounded_memoizer(maxsize: int = 128) -> callable` that:\n"
        "1. Validates that `maxsize` is a positive integer (rejecting non-integers, booleans, and non-positive numbers), raising a `ValueError` if invalid.\n"
        "2. Acts as a decorator factory returning a decorator that wraps any target callable.\n"
        "3. Caches computation results in the wrapper closure using the arguments tuple as the lookup key.\n"
        "4. Checks cache containment before evaluating the target function, tracking hit and miss telemetry accordingly.\n"
        "5. Implements a bounded FIFO eviction policy: when cache capacity reaches `maxsize` prior to inserting a new result, the oldest inserted entry is evicted.\n"
        "6. Exposes a `.cache_info()` helper method on the wrapped function that returns a dictionary containing 'hits', 'misses', 'size', and 'maxsize'.\n"
        "7. Exposes a `.cache_clear()` helper method on the wrapped function that clears the cached results and resets telemetry counters to zero.\n"
        "8. Preserves the target function's introspection metadata (such as name and docstrings)."
    ),
    "starter_code": r'''def make_bounded_memoizer(maxsize: int = 128) -> callable:
    """
    Factory creating a memoization decorator with bounded FIFO cache size and metrics.
    
    Args:
        maxsize: Maximum number of results to cache (default 128).
        
    Returns:
        Decorator function.
    """
    # TODO: Implement bounded memoizer closure with cache_info and cache_clear
    pass
''',
    "reference_solution": r'''from functools import wraps

def make_bounded_memoizer(maxsize: int = 128) -> callable:
    if not isinstance(maxsize, int) or isinstance(maxsize, bool) or maxsize <= 0:
        raise ValueError("maxsize must be a positive integer")
        
    def decorator(fn: callable) -> callable:
        if not callable(fn):
            raise TypeError("fn must be callable")
            
        cache = {}
        hits = 0
        misses = 0
        
        @wraps(fn)
        def wrapper(*args):
            nonlocal hits, misses
            key = args
            
            if key in cache:
                hits += 1
                return cache[key]
                
            misses += 1
            result = fn(*args)
            
            if len(cache) >= maxsize:
                oldest_key = next(iter(cache))
                del cache[oldest_key]
                
            cache[key] = result
            return result
            
        def cache_info():
            return {
                "hits": hits,
                "misses": misses,
                "size": len(cache),
                "maxsize": maxsize,
            }
            
        def cache_clear():
            nonlocal hits, misses
            cache.clear()
            hits = 0
            misses = 0
            
        wrapper.cache_info = cache_info
        wrapper.cache_clear = cache_clear
        return wrapper
        
    return decorator
''',
    "test_suite": r'''def run_tests(candidate_func):
    report = {"passed": True, "tests_run": 0, "errors": []}
    
    def assert_test(condition, msg):
        report["tests_run"] += 1
        if not condition:
            report["passed"] = False
            report["errors"].append(msg)
            raise AssertionError(msg)

    # Test 1: Hit and miss tracking
    call_counts = 0
    memo = candidate_func(maxsize=10)
    
    @memo
    def compute(x):
        nonlocal call_counts
        call_counts += 1
        return x * 10

    assert_test(compute(2) == 20, "Incorrect computation result")
    assert_test(compute(2) == 20, "Incorrect computation result on cache hit")
    assert_test(compute(3) == 30, "Incorrect computation result")
    
    info = compute.cache_info()
    assert_test(info["hits"] == 1, f"Expected 1 hit, got {info['hits']}")
    assert_test(info["misses"] == 2, f"Expected 2 misses, got {info['misses']}")
    assert_test(info["size"] == 2, f"Expected size 2, got {info['size']}")

    # Test 2: FIFO eviction on maxsize
    memo2 = candidate_func(maxsize=2)
    @memo2
    def square(x):
        return x * x

    square(1)  # cached: 1
    square(2)  # cached: 1, 2
    square(3)  # evicts 1, caches: 2, 3
    
    info2 = square.cache_info()
    assert_test(info2["size"] == 2, f"Cache exceeded maxsize: {info2['size']}")
    
    # Test 3: Cache clear
    compute.cache_clear()
    cleared_info = compute.cache_info()
    assert_test(cleared_info["hits"] == 0 and cleared_info["size"] == 0, "Failed to clear cache")

    return report
''',
    "hints": [
        "Maintain cache storage alongside hit and miss counters in the decorator's closure, declaring closure variables as nonlocal when modifying them in inner scopes.",
        "Check cache containment prior to evaluating the target function to return cached results and update hit or miss telemetry.",
        "When cache capacity is reached, evict the oldest inserted key relying on standard dictionary insertion order before storing new results.",
        "Attach inspection and clearance helper functions directly as callable attributes on the returned wrapper, and preserve the target function's metadata."
    ]
}

CHALLENGES = [CHALLENGE_1, CHALLENGE_2]

CURRICULUM_DATA = {
    **DAY_METADATA,
    "concept_primer": CONCEPT_PRIMER,
    "walkthrough": WALKTHROUGH,
    "challenges": CHALLENGES
}

def get_curriculum() -> Dict[str, Any]:
    return CURRICULUM_DATA
